server/auto_updater/embedding_bin_updater.py

```diff
@@ -0,0 +1,307 @@
+    )
+
+    return pca_sim_df
+
+
+def get_odds_ratio(bin_dict, background_dict, background_sum, bin_sum, cutoff_score=20):
+    """
+    This function calculates the odds ratio between
+    tokens within each bin and background token distribution.
+    Reports the log odds ratio for each bin.
+
+    Parameters:
+        bin_dict - the dictionary containing tokens within a given bin and their respective counts
+        background_dict - the dictionary containing tokens and their respective counts
+        background_sum - total number of tokens
+        bin_sum - total number of tokens in a given bin
+        cutoff_score - a threshold to remove tokens
+    """
+    # Try and filter out low count tokens to speed function up
+    filtered_bin_dict = {
+        lemma: bin_dict[lemma] for lemma in bin_dict if bin_dict[lemma] > cutoff_score
+    }
+
+    if len(filtered_bin_dict) > 0:
+        bin_dict = filtered_bin_dict
+
+    # Calculate odds ratio
+    bin_words = set(bin_dict.keys())
+    background_words = set(background_dict.keys())
+    words_to_compute = bin_words & background_words
+
+    word_odd_ratio_records = []
+    for idx, word in enumerate(words_to_compute):
+        top = float(bin_dict[word] * background_sum)
+        bottom = float(background_dict[word] * bin_sum)
+        word_odd_ratio_records.append(
+            {"lemma": word, "odds_ratio": np.log(top / bottom)}
+        )
+
+    return word_odd_ratio_records
+
+
+def update_paper_bins_stats(
+    word_bin_dict,
+    background_dict,
+    paper_embeddings_file,
+    pca_axes_file,
+    paper_landscape_json_file,
+):
+    """
+    This function performs all the updates necessary for the frontend to work.
+    It cycles through each bin and first calculates the cosine similarity between the bin
+    and each PC. Next it calculates the odds ratio for each bin given the background.
+    Lastly, it writes out all calculated statistics into a json file for the front end to use.
+
+    Parameters:
+        word_bin_dict - the dicionary containing token counts for word bins
+        background_dict - the dictionary containing token counts for all papers
+        paper_embeddings_file - the file that contains papers and their respective embeddings
+        pca_axes_file- the file containing the PC axes
+        paper_landscape_json_file - the file containing the json file the front end uses
+    """
+
+    # Grab the PCA axes and Word counts
+    # PCA
+    pca_axes_df = pd.read_csv(pca_axes_file, sep="\t")
+
+    # Read embeddings into dictionary
+    bin_centroid = defaultdict(dict)
+    if Path(paper_embeddings_file).suffix == ".xz":
+        infile = lzma.open(paper_embeddings_file, "rt")
+    else:
+        infile = open(paper_embeddings_file, "r")
+
+    reader = csv.DictReader(infile, delimiter="\t")
+
+    for line_num, line in tqdm.tqdm(enumerate(reader)):
+        if line_num == 0:
+            dim = len(line) - 2
+
+        bin_key = doc_bin_mapper[line["document"]]
```

Comment:
What is `doc_bin_mapper`?

Reply:
It's suppose to be a dictionary that has document ids and keys and the bins they are assigned as values. I forgot to pass that object along to the function, so I fixed it.